lib/grpc/client/adapters/mint/connection_process/connection_process.ex

```diff
@@ -0,0 +1,395 @@
+    window_size = get_window_size(state.conn, ref)
+    body_size = IO.iodata_length(body)
+    dequeued_state = State.update_request_stream_queue(state, queue)
+
+    cond do
+      # Do nothing, wait for server (on stream/2) to give us more window size
+      window_size == 0 -> {:noreply, state}
+      body_size > window_size -> chunk_body_and_enqueue_rest(request, dequeued_state)
+      true -> stream_body_and_reply(request, dequeued_state)
+    end
+  end
+
+  @impl true
+  def terminate(_reason, _state) do
+    :normal
+  end
+
+  defp process_response({:status, request_ref, status}, state) do
+    State.update_response_status(state, request_ref, status)
+  end
+
+  defp process_response({:headers, request_ref, headers}, state) do
+    if State.empty_headers?(state, request_ref) do
+      new_state = State.update_response_headers(state, request_ref, headers)
+
+      :ok =
+        new_state
+        |> State.stream_response_pid(request_ref)
+        |> StreamResponseProcess.consume(:headers, headers)
+
+      new_state
+    else
+      :ok =
+        state
+        |> State.stream_response_pid(request_ref)
+        |> StreamResponseProcess.consume(:trailers, headers)
+
+      state
+    end
+  end
+
+  defp process_response({:data, request_ref, new_data}, state) do
+    :ok =
+      state
+      |> State.stream_response_pid(request_ref)
+      |> StreamResponseProcess.consume(:data, new_data)
+
+    state
+  end
+
+  defp process_response({:done, request_ref}, state) do
+    :ok =
+      state
+      |> State.stream_response_pid(request_ref)
+      |> StreamResponseProcess.done()
+
+    {_ref, new_state} = State.pop_ref(state, request_ref)
+    new_state
+  end
+
+  defp chunk_body_and_enqueue_rest({request_ref, body, from}, state) do
+    {head, tail} = chunk_body(body, get_window_size(state.conn, request_ref))
+
+    case Mint.HTTP.stream_request_body(state.conn, request_ref, head) do
+      {:ok, conn} ->
+        queue = :queue.in_r({request_ref, tail, from}, state.request_stream_queue)
+
+        new_state =
+          state
+          |> State.update_conn(conn)
+          |> State.update_request_stream_queue(queue)
+
+        {:noreply, new_state}
+
+      {:error, conn, error} ->
+        if from != nil do
+          GenServer.reply(from, {:error, error})
+        else
+          :ok =
+            state
+            |> State.stream_response_pid(request_ref)
+            |> StreamResponseProcess.consume(:error, error)
+        end
+
+        {:noreply, State.update_conn(state, conn)}
+    end
+  end
+
+  defp stream_body_and_reply({request_ref, body, from}, state) do
+    send_eof? = from == nil
+
+    case stream_body(state.conn, request_ref, body, send_eof?) do
+      {:ok, conn} ->
+        if not is_nil(from) do
+          GenServer.reply(from, :ok)
+        end
+
+        check_request_stream_queue(State.update_conn(state, conn))
+
+      {:error, conn, error} ->
+        if not is_nil(from) do
```

Comment:
```suggestion
        if not send_eof? do
```

Reply:
Same question about who is expecting the reply as well

Reply:
For RPCs that has any sort of client streaming, when this [function is called](https://github.com/elixir-grpc/grpc/blob/master/lib/grpc/client/stream.ex#L91) we need to synchronously wait for that piece of request to finish it's stream before we can continue with the execution. So, I need to reply the caller process so that it can continue with the execution.